<hr/>
<h1>2. Data Preparation and Cleaning</h1>

<p>In this section, I will be preparing the datasets provided and cleaning it. This will prepare the way for Data Visualisations which will be used to answer the Research Question and fulfil the Research Objectives.</p>

<hr/>
<h2>2.1 Importing of Packages</h2>

<p>Before conducting Data Preparation and Cleaning, I will firstly import the required modules which are Numpy, Pandas and Matplot. These modules are required to perform the Data Analytics with the datasets.</p>

In [16]:
# --------- Importing NumPy as Alias 'np' --------- #
import numpy as np

# --------- Importing Pandas as Alias 'pd' --------- #
import pandas as pd

# --------- Importing Matplotlib as Alias 'plt' --------- #
import matplotlib.pyplot as plt

# --------- Confirmation of Modules Importation --------- #
print("Modules Successfully Imported")

Modules Successfully Imported


<p>With the output of 'Modules Successfully Imported', I am able to verify that the modules have been imported and I am able to begin Data Preparation and Cleaning</p>

<hr/>
<h2>2.2 Data Overview</h2>

<p>To make sense of the dataset provided, I will firstly need to view each dataset individually and verify the individual contents within the dataframe, its data's presence and also the possibility of merging dataframes.</p>

<hr/>
<h3>2.2.1 First Dataset</h3>

<p>The Data Overview will be done on the first dataset in this section. Some operations which will be perfromed are, showing dataset's content, checking length of dataset, checking for missing values, etc.</p>

<hr/>
<h4>2.2.1.1 First Dataset's Content</h4>

<p>I will be importing the first dataset and subsequently view the first 5 data columns of the dataset.</p>

In [24]:
# ----------- Importing Dataset 1 & Assigning to Alias 'df1' ----------- #
df1 = pd.read_csv('./dataset/postings1.csv')

# ----------- Viewing df1's First 5 Data Rows ----------- #
df1.head(5)

,job_id,company_name,title
0,921716,Corcoran Sawyer Smith,Marketing Coordinator
1,1829192,NaN,Mental Health Therapist/Counselor
2,10998357,The National Exemplar,Assitant Restaurant Manager
3,23221523,"Abrams Fensterman, LLP",Senior Elder Law / Trusts and Estates Associat...
4,35982263,NaN,Service Technician


<p>From the dataframe above, I am able to see that there is 1 unique identifier for each of the data rows. I am also able to see that the dataframe contains a total of 3 data columns (inclusive of the UIDs). I can also conclude that there are a total of 2 cateogircal data columns. A detailed report is listed below.</p>
<br/>
<p><strong><u>Unique Identifier</u></strong></p>
<ul>
    <li>job_id</li>
</ul>
<p><strong><u>Categorical Data</u></strong></p>
<ul>
    <li>company_name</li>
    <li>title</li>
</ul>
<p><strong><u>Numerical Data</u></strong></p>
<ul>
    <li>NONE</li>
</ul>
<p><strong><u>Textual Data</u></strong></p>
<ul>
    <li>NONE</li>
</ul>
<br/>
<p>With the observations stated above, I will be moving on to see the total data rows within the dataset.</p>

<hr/>
<h4>2.2.1.2 First Dataset's Shape</h4>

<p>I will be viewing the shape of the dataframe to see how many data columns there are to fully comprehend the size and scale of the dataset provided to me. This will also allow for cross-referencing where I am able to see if all the dataset matches each other in terms for size and scale.</p>

In [28]:
# ----------- Printing df1's Shape ----------- #
df1.shape

(123849, 3)

<p>From the output above, I am able to verify that there is a total of 123849 data rows within the first dataset. This shows that I am working with a large dataset with multiple data rows. This also means that the Data Analysis done will be more accurate and relevant considering the wide range of data and information within the dataset. This will make the project much more applicable to the real-life context of the job market.</p>

<hr/>
<h4>2.2.1.3 First Dataset's Missing Data</h4>

<p>I will be checking for missing data within the dataset. This will allow me to plan ahead on whether there is a need to impute missing data to fill up the gaps within these data columns.</p>

In [32]:
# ----------- Placing Missing Values of df1 Into A Dataframe ----------- #
df1MissingData = pd.DataFrame({'Missing Values': df1.isnull().sum().values}, index = df1.columns)

# ----------- Function for Colour Coding ----------- #
def colour_code(values):
    return 'background-color: green' if values == 0 else 'background-color: red'

# ----------- Calling Function for Colour Coding ----------- #
df1MissingData = df1MissingData.style.map(colour_code, subset = ['Missing Values'])

# ----------- Displaying Missing Values ----------- #
df1MissingData

,Missing Values
job_id,0
company_name,1719
title,0


<p>From the output above, I am able to make the following observations.</p>
<br/>
<p><strong><u>Data Column with Missing Data</u></strong></p>
<ul>
    <li>company_name (1719 Missing Data)</li>
</ul>
<br/>
<p>From the observations above, I can conclude that some data columns may require me to impute the missing data.</p>

<hr/>
<h3>2.2.2 Second Dataset</h3>

<p>The Data Overview will be done on the second dataset in this section. Some operations which will be perfromed are, showing dataset's content, checking length of dataset, checking for missing values, etc.</p>

<hr/>
<h4>2.2.2.1 Second Dataset's Content</h4>

<p>I will be importing the second dataset and subsequently view the first 5 data columns of the dataset.</p>

In [38]:
# ----------- Importing Dataset 2 & Assigning to Alias 'df2' ----------- #
df2 = pd.read_csv('./dataset/postings2.csv')

# ----------- Viewing df2's First 5 Data Rows ----------- #
df2.head(5)

,job_id,description,max_salary
0,921716,Job descriptionA leading real estate firm in N...,20.0
1,1829192,"At Aspen Therapy and Wellness , we are committ...",50.0
2,10998357,The National Exemplar is accepting application...,65000.0
3,23221523,Senior Associate Attorney - Elder Law / Trusts...,175000.0
4,35982263,Looking for HVAC service tech with experience ...,80000.0


<p>From the dataframe above, I am able to see that there is 1 unique identifier for each of the data rows. I am also able to see that the dataframe contains a total of 3 data columns (inclusive of the UIDs). I can also conclude that there are a total of 1 textual data column and 1 numerical data column. A detailed report is listed below.</p>
<br/>
<p><strong><u>Unique Identifier</u></strong></p>
<ul>
    <li>job_id</li>
</ul>
<p><strong><u>Categorical Data</u></strong></p>
<ul>
    <li>NONE</li>
</ul>
<p><strong><u>Numerical Data</u></strong></p>
<ul>
    <li>max_salary</li>
</ul>
<p><strong><u>Textual Data</u></strong></p>
<ul>
    <li>description</li>
</ul>
<br/>
<p>With the observations stated above, I will be moving on to see the total data rows within the dataset.</p>

<hr/>
<h4>2.2.2.2 Second Dataset's Shape</h4>

<p>I will be viewing the shape of the dataframe to see how many data columns there are to fully comprehend the size and scale of the dataset provided to me. This will also allow for cross-referencing where I am able to see if all the dataset matches each other in terms for size and scale.</p>

In [42]:
# ----------- Printing df2's Shape ----------- #
df2.shape

(123849, 3)

<p>From the output above, I am able to verify that there is a total of 123849 data rows within the first dataset. This shows that I am working with a large dataset with multiple data rows. This also means that the Data Analysis done will be more accurate and relevant considering the wide range of data and information within the dataset. This will make the project much more applicable to the real-life context of the job market.</p>

<hr/>
<h4>2.2.2.3 Second Dataset's Missing Data</h4>

<p>I will be checking for missing data within the dataset. This will allow me to plan ahead on whether there is a need to impute missing data to fill up the gaps within these data columns.</p>

In [46]:
# ----------- Placing Missing Values of df2 Into A Dataframe ----------- #
df2MissingData = pd.DataFrame({'Missing Values': df2.isnull().sum().values}, index = df2.columns)

# ----------- Function for Colour Coding ----------- #
def colour_code(values):
    return 'background-color: green' if values == 0 else 'background-color: red'

# ----------- Calling Function for Colour Coding ----------- #
df2MissingData = df2MissingData.style.map(colour_code, subset = ['Missing Values'])

# ----------- Displaying Missing Values ----------- #
df2MissingData

,Missing Values
job_id,0
description,7
max_salary,94056


<p>From the output above, I am able to make the following observations.</p>
<br/>
<p><strong><u>Data Column with Missing Data</u></strong></p>
<ul>
    <li>description (7 Missing Data)</li>
    <li>max_salary (94056 Missing Data)</li>
</ul>
<br/>
<p>From the observations above, I can conclude that some data columns may require me to impute the missing data.</p>

<hr/>
<h3>2.2.3 Third Dataset</h3>

<p>The Data Overview will be done on the third dataset in this section. Some operations which will be perfromed are, showing dataset's content, checking length of dataset, checking for missing values, etc.</p>

<hr/>
<h4>2.2.3.1 Third Dataset's Content</h4>

<p>I will be importing the third dataset and subsequently view the first 5 data columns of the dataset.</p>

In [52]:
# ----------- Importing Dataset 3 & Assigning to Alias 'df3' ----------- #
df3 = pd.read_csv('./dataset/postings3.csv')

# ----------- Viewing df3's First 5 Data Rows ----------- #
df3.head(5)

,job_id,pay_period,location,company_id,views,med_salary,min_salary
0,921716,HOURLY,"Princeton, NJ",2774458.0,20.0,NaN,17.0
1,1829192,HOURLY,"Fort Collins, CO",NaN,1.0,NaN,30.0
2,10998357,YEARLY,"Cincinnati, OH",64896719.0,8.0,NaN,45000.0
3,23221523,YEARLY,"New Hyde Park, NY",766262.0,16.0,NaN,140000.0
4,35982263,YEARLY,"Burlington, IA",NaN,3.0,NaN,60000.0


<p>From the dataframe above, I am able to see that there is 2 unique identifier for each of the data rows. I am also able to see that the dataframe contains a total of 7 data columns (inclusive of the UIDs). I can also conclude that there are a total of 2 categorical data column and 3 numerical data column. A detailed report is listed below.</p>
<br/>
<p><strong><u>Unique Identifier</u></strong></p>
<ul>
    <li>job_id</li>
    <li>company_id</li>
</ul>
<p><strong><u>Categorical Data</u></strong></p>
<ul>
    <li>pay_period</li>
    <li>location</li>
</ul>
<p><strong><u>Numerical Data</u></strong></p>
<ul>
    <li>views</li>
    <li>med_salary</li>
    <li>min_salary</li>
</ul>
<br/>
<p>With the observations stated above, I will be moving on to see the total data rows within the dataset.</p>

<hr/>
<h4>2.2.3.2 Third Dataset's Shape</h4>

<p>I will be viewing the shape of the dataframe to see how many data columns there are to fully comprehend the size and scale of the dataset provided to me. This will also allow for cross-referencing where I am able to see if all the dataset matches each other in terms for size and scale.</p>

In [56]:
# ----------- Printing df3's Shape ----------- #
df3.shape

(123849, 7)

<p>From the output above, I am able to verify that there is a total of 123849 data rows within the first dataset. This shows that I am working with a large dataset with multiple data rows. This also means that the Data Analysis done will be more accurate and relevant considering the wide range of data and information within the dataset. This will make the project much more applicable to the real-life context of the job market.</p>

<hr/>
<h4>2.2.3.3 Third Dataset's Missing Data</h4>

<p>I will be checking for missing data within the dataset. This will allow me to plan ahead on whether there is a need to impute missing data to fill up the gaps within these data columns.</p>

In [60]:
# ----------- Placing Missing Values of df3 Into A Dataframe ----------- #
df3MissingData = pd.DataFrame({'Missing Values': df3.isnull().sum().values}, index = df3.columns)

# ----------- Function for Colour Coding ----------- #
def colour_code(values):
    return 'background-color: green' if values == 0 else 'background-color: red'

# ----------- Calling Function for Colour Coding ----------- #
df3MissingData = df3MissingData.style.map(colour_code, subset = ['Missing Values'])

# ----------- Displaying Missing Values ----------- #
df3MissingData

,Missing Values
job_id,0
pay_period,87776
location,0
company_id,1717
views,1689
med_salary,117569
min_salary,94056


<p>From the output above, I am able to make the following observations.</p>
<br/>
<p><strong><u>Data Column with Missing Data</u></strong></p>
<ul>
    <li>pay_period (87776 Missing Data)</li>
    <li>company_id (1717 Missing Data)</li>
    <li>views (1689 Missing Data)</li>
    <li>med_salary (117569 Missing Data)</li>
    <li>min_salary (94056 Missing Data)</li>
</ul>
<br/>
<p>From the observations above, I can conclude that some data columns may require me to impute the missing data.</p>

<hr/>
<h3>2.2.4 Fourth Dataset</h3>

<p>The Data Overview will be done on the fourth dataset in this section. Some operations which will be perfromed are, showing dataset's content, checking length of dataset, checking for missing values, etc.</p>

<hr/>
<h4>2.2.4.1 Fourth Dataset's Content</h4>

<p>I will be importing the fourth dataset and subsequently view the first 5 data columns of the dataset.</p>

In [66]:
# ----------- Importing Dataset 4 & Assigning to Alias 'df4' ----------- #
df4 = pd.read_csv('./dataset/postings4.csv')

# ----------- Viewing df4's First 5 Data Rows ----------- #
df4.head(5)

,job_id,formatted_work_type,applies,original_listed_time,remote_allowed,job_posting_url,application_url,application_type,expiry,closed_time,formatted_experience_level,skills_desc,listed_time,posting_domain,sponsored,work_type,currency,compensation_type
0,921716,Full-time,2.0,1.713400e+12,NaN,https://www.linkedin.com/jobs/view/921716/?trk...,NaN,ComplexOnsiteApply,1.715990e+12,NaN,NaN,Requirements: \n\nWe are seeking a College or ...,1.713400e+12,NaN,0,FULL_TIME,USD,BASE_SALARY
1,1829192,Full-time,NaN,1.712860e+12,NaN,https://www.linkedin.com/jobs/view/1829192/?tr...,NaN,ComplexOnsiteApply,1.715450e+12,NaN,NaN,NaN,1.712860e+12,NaN,0,FULL_TIME,USD,BASE_SALARY
2,10998357,Full-time,NaN,1.713280e+12,NaN,https://www.linkedin.com/jobs/view/10998357/?t...,NaN,ComplexOnsiteApply,1.715870e+12,NaN,NaN,We are currently accepting resumes for FOH - A...,1.713280e+12,NaN,0,FULL_TIME,USD,BASE_SALARY
3,23221523,Full-time,NaN,1.712900e+12,NaN,https://www.linkedin.com/jobs/view/23221523/?t...,NaN,ComplexOnsiteApply,1.715490e+12,NaN,NaN,This position requires a baseline understandin...,1.712900e+12,NaN,0,FULL_TIME,USD,BASE_SALARY
4,35982263,Full-time,NaN,1.713450e+12,NaN,https://www.linkedin.com/jobs/view/35982263/?t...,NaN,ComplexOnsiteApply,1.716040e+12,NaN,NaN,NaN,1.713450e+12,NaN,0,FULL_TIME,USD,BASE_SALARY


<p>From the dataframe above, I am able to see that there is 3 unique identifiers (UIDs) within the dataset. I am also able to see that the dataframe contains a total of 18 data columns (inclusive of UIDs). I can also conclude that there are a total of 10 categorical data column and 4 numerical data column (exclusive of UID) and 1 Textual Data. A detailed report is listed below.</p>
<br/>
<p><strong><u>Unique Identifier</u></strong></p>
<ul>
    <li>job_id</li>
    <li>job_posting_url</li>
    <li>application_url</li>
</ul>
<p><strong><u>Categorical Data</u></strong></p>
<ul>
    <li>formatted_work_type</li>
    <li>remote_allowed (Dummy Variable)</li>
    <li>application_type</li>
    <li>formatted_experience_level	</li>
    <li>skills_desc</li>
    <li>posting_domain</li>
    <li>sponsored</li>
    <li>work_type</li>
    <li>currency</li>
    <li>compensation_type
</li>
</ul>
<p><strong><u>Numerical Data</u></strong></p>
<ul>
    <li>applies</li>
    <li>original_listed_time</li>
    <li>expiry</li>
    <li>closed_time</li>
</ul>
<p><strong><u>Textual Data</u></strong></p>
<ul>
    <li>skills_desc</li>
</ul>
<br/>
<p>With the observations stated above, I will be moving on to see the total data rows within the dataset.</p>

<hr/>
<h4>2.2.4.2 Fourth Dataset's Shape</h4>

<p>I will be viewing the shape of the dataframe to see how many data columns there are to fully comprehend the size and scale of the dataset provided to me. This will also allow for cross-referencing where I am able to see if all the dataset matches each other in terms for size and scale.</p>

In [70]:
# ----------- Printing df4's Shape ----------- #
df4.shape

(123849, 18)

<p>From the output above, I am able to verify that there is a total of 123849 data rows within the first dataset. This shows that I am working with a large dataset with multiple data rows. This also means that the Data Analysis done will be more accurate and relevant considering the wide range of data and information within the dataset. This will make the project much more applicable to the real-life context of the job market.</p>

<hr/>
<h4>2.2.4.3 Fourth Dataset's Missing Data</h4>

<p>I will be checking for missing data within the dataset. This will allow me to plan ahead on whether there is a need to impute missing data to fill up the gaps within these data columns.</p>

In [74]:
# ----------- Placing Missing Values of df4 Into A Dataframe ----------- #
df4MissingData = pd.DataFrame({'Missing Values': df4.isnull().sum().values}, index = df4.columns)

# ----------- Function for Colour Coding ----------- #
def colour_code(values):
    return 'background-color: green' if values == 0 else 'background-color: red'

# ----------- Calling Function for Colour Coding ----------- #
df4MissingData = df4MissingData.style.map(colour_code, subset = ['Missing Values'])

# ----------- Displaying Missing Values ----------- #
df4MissingData

,Missing Values
job_id,0
formatted_work_type,0
applies,100529
original_listed_time,0
remote_allowed,108603
job_posting_url,0
application_url,36665
application_type,0
expiry,0
closed_time,122776


<p>From the output above, I am able to make the following observations.</p>
<br/>
<p><strong><u>Data Column with Missing Data</u></strong></p>
<ul>
    <li>applies (100529 Missing Data)</li>
    <li>remote_allowed (108603 Missing Data)</li>
    <li>application_url (36665 Missing Data)</li>
    <li>closed_time (122776 Missing Data)</li>
    <li>formatted_experience_level (29409 Missing Data)</li>
    <li>skills_desc (121410 Missing Data)</li>
    <li>posting_domain (39968 Missing Data)</li>
    <li>currency (87776 Missing Data)</li>
    <li>compensation_type (87776 Missing Data)</li>
</ul>
<br/>
<p>From the observations above, I can conclude that some data columns may require me to impute the missing data.</p>

<hr/>
<h3>2.2.5 Fifth Dataset</h3>

<p>The Data Overview will be done on the fifth dataset in this section. Some operations which will be perfromed are, showing dataset's content, checking length of dataset, checking for missing values, etc.</p>

<hr/>
<h4>2.2.5.1 Fifth Dataset's Content</h4>

<p>I will be importing the fifth dataset and subsequently view the first 5 data rows of the dataset.</p>

In [80]:
# ----------- Importing Dataset 5 & Assigning to Alias 'df5' ----------- #
df5 = pd.read_excel('./dataset/postings5.xlsx')

# ----------- Viewing df5's First 5 Data Rows ----------- #
df5.head(5)

,job_id,company_name,title,Easter Egg,Unimportant Column
0,921716,Corcoran Sawyer Smith,Marketing Coordinator,"In order to score well for this module, you sh...",1.0
1,1829192,NaN,Mental Health Therapist/Counselor,"In order to score well for this module, you sh...",1.0
2,10998357,The National Exemplar,Assitant Restaurant Manager,"In order to score well for this module, you sh...",1.0
3,23221523,"Abrams Fensterman, LLP",Senior Elder Law / Trusts and Estates Associat...,"In order to score well for this module, you sh...",1.0
4,35982263,NaN,Service Technician,"In order to score well for this module, you sh...",1.0


<p>From the dataframe above, I am able to see that there is 1 unique identifier (UID) within the dataset. I am also able to see that the dataframe contains a total of 5 data columns (inclusive of UIDs). I can also conclude that there are a total of 2 categorical data column and 2 miscellaneous data columns (exclusive of UID). A detailed report is listed below.</p>
<br/>
<p><strong><u>Unique Identifier</u></strong></p>
<ul>
    <li>job_id</li>
</ul>
<p><strong><u>Categorical Data</u></strong></p>
<ul>
    <li>company_name</li>
    <li>title</li>
</ul>
<p><strong><u>Numerical Data</u></strong></p>
<ul>
    <li>NONE</li>
</ul>
<p><strong><u>Textual Data</u></strong></p>
<ul>
    <li>NONE</li>
</ul>
<p><strong><u>Miscellaneous Data</u></strong></p>
<ul>
    <li>Easter Egg</li>
    <li>Unimportant Column</li>
</ul>
<br/>
<p>With the observations stated above, I will be moving on to see the total data rows within the dataset.</p>

<hr/>
<h4>2.2.5.2 Fifth Dataset's Shape</h4>

<p>I will be viewing the shape of the dataframe to see how many data columns there are to fully comprehend the size and scale of the dataset provided to me. This will also allow for cross-referencing where I am able to see if all the dataset matches each other in terms for size and scale.</p>

In [84]:
# ----------- Printing df5's Shape ----------- #
df5.shape

(123849, 5)

<p>From the output above, I am able to verify that there is a total of 123849 data rows within the first dataset. This shows that I am working with a large dataset with multiple data rows. This also means that the Data Analysis done will be more accurate and relevant considering the wide range of data and information within the dataset. This will make the project much more applicable to the real-life context of the job market.</p>

<hr/>
<h4>2.2.5.3 Fifth Dataset's Missing Data</h4>

<p>I will be checking for missing data within the dataset. This will allow me to plan ahead on whether there is a need to impute missing data to fill up the gaps within these data columns.</p>

In [88]:
# ----------- Placing Missing Values of df5 Into A Dataframe ----------- #
df5MissingData = pd.DataFrame({'Missing Values': df5.isnull().sum().values}, index = df5.columns)

# ----------- Function for Colour Coding ----------- #
def colour_code(values):
    return 'background-color: green' if values == 0 else 'background-color: red'

# ----------- Calling Function for Colour Coding ----------- #
df5MissingData = df5MissingData.style.map(colour_code, subset = ['Missing Values'])

# ----------- Displaying Missing Values ----------- #
df5MissingData

,Missing Values
job_id,0
company_name,1719
title,0
Easter Egg,115041
Unimportant Column,119672


<p>From the output above, I am able to make the following observations.</p>
<br/>
<p><strong><u>Data Column with Missing Data</u></strong></p>
<ul>
    <li>company_name (1719 Missing Data)</li>
    <li>Easter Egg (115041 Missing Data)</li>
    <li>Unimportant Column (119672 Missing Data)</li>
</ul>
<br/>
<p>From the observations above, I can conclude that some data columns may require me to impute the missing data. However, the 'Easter Egg' column and 'Unimportant Column' are miscellaneous, thus no imputation is required.</p>

<hr/>
<h2>2.3 Merging and Filtering Dataset</h2>

<p>From the previous sub-section, I identified that all dataset has the same amount of data rows (123849 data rows). I also indentified that all dataset has a common UID. This allows us to merge all the datasets together and subsequently filter out data columns that are not needed before imputing the missing data if required.</p>

In [92]:
# ----------- Merging of All Dataframes ----------- #
merged_df = (df1.merge(df2, on = 'job_id', how = 'inner').merge(df3, on = 'job_id', how = 'inner').merge(df4, on = 'job_id', how = 'inner').merge(df5, on = 'job_id', how = 'inner'))

# ----------- Setting Index to Job ID ----------- #
merged_df = merged_df.set_index('job_id', drop = True)

# ----------- Identification of Irrelevant Data Columns ----------- #
irrelevantDataColumns = [
    'job_posting_url', 'sponsored', 'posting_domain',
    'application_url', 'description', 'original_listed_time',
    'skills_desc', 'closed_time',
    'company_id', 'currency', 'compensation_type',
    'work_type', 'med_salary', 'Easter Egg',
    'Unimportant Column', 'title_y', 'company_name_y'
]

# ----------- Dropping of Irrelevant Data Columns ----------- #
merged_df = merged_df.drop(columns = irrelevantDataColumns)

# ----------- Display of DataFrame ----------- #
merged_df.head(10)

,company_name_x,title_x,max_salary,pay_period,location,views,min_salary,formatted_work_type,applies,remote_allowed,application_type,expiry,formatted_experience_level,listed_time
job_id,,,,,,,,,,,,,,
921716,Corcoran Sawyer Smith,Marketing Coordinator,20.0,HOURLY,"Princeton, NJ",20.0,17.0,Full-time,2.0,NaN,ComplexOnsiteApply,1.715990e+12,NaN,1.713400e+12
1829192,NaN,Mental Health Therapist/Counselor,50.0,HOURLY,"Fort Collins, CO",1.0,30.0,Full-time,NaN,NaN,ComplexOnsiteApply,1.715450e+12,NaN,1.712860e+12
10998357,The National Exemplar,Assitant Restaurant Manager,65000.0,YEARLY,"Cincinnati, OH",8.0,45000.0,Full-time,NaN,NaN,ComplexOnsiteApply,1.715870e+12,NaN,1.713280e+12
23221523,"Abrams Fensterman, LLP",Senior Elder Law / Trusts and Estates Associat...,175000.0,YEARLY,"New Hyde Park, NY",16.0,140000.0,Full-time,NaN,NaN,ComplexOnsiteApply,1.715490e+12,NaN,1.712900e+12
35982263,NaN,Service Technician,80000.0,YEARLY,"Burlington, IA",3.0,60000.0,Full-time,NaN,NaN,ComplexOnsiteApply,1.716040e+12,NaN,1.713450e+12
91700727,Downtown Raleigh Alliance,Economic Development and Planning Intern,20.0,HOURLY,"Raleigh, NC",9.0,14.0,Internship,4.0,NaN,ComplexOnsiteApply,1.716050e+12,NaN,1.713460e+12
103254301,Raw Cereal,Producer,300000.0,YEARLY,United States,7.0,60000.0,Contract,1.0,1.0,SimpleOnsiteApply,1.715450e+12,NaN,1.712860e+12
112576855,NaN,Building Engineer,120000.0,YEARLY,"San Francisco, CA",2.0,90000.0,Full-time,NaN,NaN,ComplexOnsiteApply,1.728000e+12,NaN,1.712440e+12
1218575,Children's Nebraska,Respiratory Therapist,NaN,NaN,"Omaha, NE",3.0,NaN,Full-time,NaN,NaN,OffsiteApply,1.714940e+12,NaN,1.712350e+12


<p>From the output of the dataframe above, I can verify that the data columns that were deemed irrelevant have been dropped and all the dataframes have been successfully merged together. The reason for dropping each data columns have been listed below.</p>
<br/>
<p><strong><u>job_posting_url</u></strong></p>
<ul>
    <li>A unique identifier that is different for all job postings and serves no purpose in Data Analytics.</li>
</ul>
<p><strong><u>sponsored</u></strong></p>
<ul>
    <li>Made up of all 0s meaning every job posting is nor sponsored. Therefore, it serves no purpose in Data Analytics as it is common for all job postings.</li>
</ul>
<p><strong><u>posting_domain</u></strong></p>
<ul>
    <li>Redundant data as it does not assist in answering the research question.</li>
</ul>
<p><strong><u>application_url</u></strong></p>
<ul>
    <li>A unique identifier that is different for all job postings / missing for some job postings. Therefore, it serves no purpose in Data Analytics.</li>
</ul>
<p><strong><u>description</u></strong></p>
<ul>
    <li>A textual data that serves no analytical purpose.</li>
</ul>
<p><strong><u>original_listed_time</u></strong></p>
<ul>
    <li>Redundant as 'listed_time' is available.</li>
</ul>
<p><strong><u>skills_desc</u></strong></p>
<ul>
    <li>A textual data that serves no analytical purpose.</li>
</ul>
<p><strong><u>closed_time</u></strong></p>
<ul>
    <li>Redundant as 'expiry' can be used for duration calculation</li>
</ul>
<p><strong><u>company_id</u></strong></p>
<ul>
    <li>Redundant as 'company_name' can be used instead.</li>
</ul>
<p><strong><u>currency</u></strong></p>
<ul>
    <li>Serves no analytical purpose as all the data fields are either 'USD' or missing.</li>
</ul>
<p><strong><u>compensation_type</u></strong></p>
<ul>
    <li>Redundant as most data fields are either 'BASE SALARY' or missing data. The 'min_salary' and 'max_salary' are also a representation of the compensation type. Therefore, the data column serves no purpose.</li>
</ul>
<p><strong><u>work_type</u></strong></p>
<ul>
    <li>Dropped due to duplication by 'formatted_work_type'.</li>
</ul>
<p><strong><u>med_salary</u></strong></p>
<ul>
    <li>Dropped due to mostly being missing values and the presence of 'min_salary' and 'max_salary' has already served the purpose of 'med_salary'.</li>
</ul>
<p><strong><u>Easter Egg</u></strong></p>
<ul>
    <li>Dropped due to it being redundant and serving no analytical purpose.</li>
</ul>
<p><strong><u>Unimportant Column</u></strong></p>
<ul>
    <li>Dropped due to it being redundant and serving no analytical purpose.</li>
</ul>
<p><strong><u>title_y</u></strong></p>
<ul>
    <li>Dropped due to duplication of 'title_x'.</li>
</ul>
<p><strong><u>comapny_name_y</u></strong></p>
<ul>
    <li>Dropped due to duplication of 'comapny_name_x'.</li>
</ul>
<br/>
<p>With all the redundant data columns dropped, I will be moving on to imputing the missing data.</p>

<hr/>
<h2>2.4 Imputing Missing Data</h2>

<p>From the Data Overview, I found that many data columns has missing data. Therefore, in the sub-section, I will be addressing these missing data through imputations. The methods of imputation will be different for each of the data columns as some data columns contains numerical data while others contain categorical data.</p>

<hr/>
<h3>2.4.1 Identifying Missing Data and Imputation Strategy</h3>

<p>Before imputing the data, I will identify the data columns with missing data one last time and subsequently, decide which imputation strategy will be most suitable in regards to the data columns.</p>

In [98]:
# ----------- Renaming of Data Columns ----------- #
merged_df = merged_df.rename(columns={'company_name_x': 'company_name', 'title_x': 'title'})

# ----------- Placing Missing Values of the Merged Dataframe Into A Dataframe ----------- #
mergedDfMissingData = pd.DataFrame({'Missing Values': merged_df.isnull().sum().values}, index = merged_df.columns)

# ----------- Function for Colour Coding ----------- #
def colour_code(val):
    if val > 0:
        return 'background-color: red; color: white'
    else:
        return 'background-color: green; color: white'

# ----------- Calling Function for Colour Coding ----------- #
mergedDfMissingData = mergedDfMissingData.style.map(colour_code, subset = ['Missing Values'])

# ----------- Displaying Missing Values----------- #
mergedDfMissingData

,Missing Values
company_name,1719
title,0
max_salary,94056
pay_period,87776
location,0
views,1689
min_salary,94056
formatted_work_type,0
applies,100529
remote_allowed,108603


<p>With reference to the output above, I have made an action report on how I shall proceed with the imputation. The action report is listed below.</p>
<br/>
<p><strong><u>company_name</u></strong></p>
<ul>
    <li>Imputation Strategy: String Filling</li>
    <li>Reason: Company name cannot be imputed. Therefore, a placeholder will be used.</li>
</ul>
<p><strong><u>max_salary</u></strong></p>
<ul>
    <li>Imputation Strategy: Group Interpolation with Experience Level and Pay Period</li>
    <li>Reason: Due to potential outliers and large number of Missing Data and requiring a realistic value.</li>
</ul>
<p><strong><u>pay_period</u></strong></p>
<ul>
    <li>Imputation Strategy: Mode Imputation</li>
    <li>Reason: Mode will ensure that the most frequent value will be used.</li>
</ul>
<p><strong><u>views</u></strong></p>
<ul>
    <li>Imputation Strategy: Integer Filling</li>
    <li>Reason: Views with missing data most likely refers to no views. Therefore, a Integer Fill of 0 will be used.</li>
</ul>
<p><strong><u>min_salary</u></strong></p>
<ul>
    <li>Imputation Strategy: Group Interpolation with Experience Level and Pay Period</li>
    <li>Reason: Due to potential outliers and large number of Missing Data and requiring a realistic value.</li>
</ul>
<p><strong><u>applies</u></strong></p>
<ul>
    <li>Imputation Strategy: Integer Filling</li>
    <li>Reason: Applies with missing data likely means that there is no application. Therefore a Integer Fill of 0 will be used.</li>
</ul>
<p><strong><u>remote_allowed</u></strong></p>
<ul>
    <li>Imputation Strategy: Integer Filling</li>
    <li>Reason: remote_allowed is a dummy variable consisting of mainly 1s. Therefore, bu having a missing data, it most likely means remote working is not allowed, Thus Integer Filling of '0' will be used.</li>
</ul>
<p><strong><u>formatted_experience_level</u></strong></p>
<ul>
    <li>Imputation Strategy: Mode Imputation</li>
    <li>Reason: Mode will ensure that the most frequent value will be used.</li>
</ul>
<br/>
<p>With the action plan listed above, I will proceed with the imputation in the next cell.</p>

<hr/>
<h3>2.4.2 Imputation of Missing Data</h3>

<p>Using the identified strategy listed in the previous section, I will be using the identified strategy on the data column with missing data.</p>

In [102]:
# ----------- String Filling ----------- #
merged_df['company_name'] = merged_df['company_name'].fillna('Unknown')

# ----------- Integer Filling ----------- #
merged_df['views'] = merged_df['views'].fillna(0)
merged_df['applies'] = merged_df['applies'].fillna(0)
merged_df['remote_allowed'] = merged_df['remote_allowed'].fillna(0)

# ----------- Mode Imputation ----------- #
merged_df['pay_period'] = merged_df['pay_period'].fillna(merged_df['pay_period'].mode()[0])
merged_df['formatted_experience_level'] = merged_df['formatted_experience_level'].fillna(
    merged_df['formatted_experience_level'].mode()[0]
)

# ----------- Function for Interpolation by Groups ----------- #
def interpolate_salary_by_group(df, column):
    # ----------- Sorting of Data ----------- #
    df = df.sort_values(by = ['pay_period', 'formatted_experience_level', 'listed_time'])
    
    # ----------- Interpolation Process ----------- #
    df[column] = df.groupby(['pay_period', 'formatted_experience_level'])[column] \
                   .transform(lambda x: x.interpolate(method = 'linear', limit_direction = 'both'))
    
    # ----------- Returning of Final Results ----------- #
    return df

# ----------- Interpolation for max_salary and min_salary ----------- #
merged_df = interpolate_salary_by_group(merged_df, 'max_salary')
merged_df = interpolate_salary_by_group(merged_df, 'min_salary')

# ----------- Placing Missing Values of the Merged Dataframe Into A Dataframe ----------- #
mergedDfMissingData = pd.DataFrame({'Missing Values': merged_df.isnull().sum().values}, index = merged_df.columns)

# ----------- Function for Colour Coding ----------- #
def colour_code(val):
    if val > 0:
        return 'background-color: red; color: white'
    else:
        return 'background-color: green; color: white'

# ----------- Calling Function for Colour Coding ----------- #
mergedDfMissingData = mergedDfMissingData.style.map(colour_code, subset = ['Missing Values'])

# ----------- Displaying Missing Values----------- #
mergedDfMissingData

,Missing Values
company_name,0
title,0
max_salary,0
pay_period,0
location,0
views,0
min_salary,0
formatted_work_type,0
applies,0
remote_allowed,0


<p>With the output of the dataframe above, I can verify that all data columns have been successfully imputed and I can now proceed with Handling Duplicates.</p>

<hr/>
<h2>2.5 Handling Duplicated Data</h2>

<p>In this sub-section, I will be identifying and removing duplicated data within my merged dataframe. Duplicated can have a negative effect on the Data Analytics which I will be performing. Therefore, it is of the utmost importance that duplicated data be identified and addressed.</p>

<hr/>
<h3>2.5.1 Identifying Duplicated Data</h3>

<p>Before being able to address duplicated data, I have to verify if there are duplicated data in the first place. In the cell below, I will be checking for duplicated data and based on the output, one of the two action plan will occur.</p>
<br/>
<span style="color:red"><p><strong><u>Duplicated Data Identified</u></strong></p></span>
<ul>
    <li>Proceed to remove the duplicated data.</li>
</ul>
<span style="color:green"><p><strong><u>No Duplicated Data Identified</u></strong></p></span>
<ul>
    <li>Proceed with next sub-section on Fixing Data Types.</li>
</ul>
<br/>
<p>With the action plan listed above, I will be proceeding to check for duplicated data.</p>

In [108]:
# ----------- Calculate Total Duplicated Rows ----------- #
duplicatesCount = merged_df.duplicated().sum()

# ----------- Create DataFrame to Store Result ----------- #
duplicateSummary = pd.DataFrame({'Count': [duplicatesCount]}, index = ['Duplicated Rows'])

# ----------- Function for Colour Coding ----------- #
def colour_code(val):
    if val > 0:
        return 'background-color: red; color: white'
    else:
        return 'background-color: green; color: white'

# ----------- Apply Colour Coding to the 'Count' Column ----------- #
duplicateSummary = duplicateSummary.style.map(colour_code, subset=['Count'])

# ----------- Display the Result ----------- #
duplicateSummary

,Count
Duplicated Rows,77


<p>From the output above, I can infer that there is a total of '77' duplicated data rows. Thus there is <span style="color:red"><strong><u>Duplicated Data Identified</u></strong></span>. As stated in the action plan, I will proceed with the removal of duplicated data.
</p>

<hr/>
<h3>2.5.2 Removing Duplicated Data</h3>

<p>With the presence of duplicated data identified and confirmed, it is essential to remove it as duplicated data can distort data analysis causing any analysis done to be inaccurate and unreliable.</p>

In [112]:
# ----------- Removing Duplicated Data ----------- #
merged_df = merged_df.drop_duplicates()

# ----------- Calculate Total Duplicated Rows ----------- #
duplicatesCount = merged_df.duplicated().sum()

# ----------- Create DataFrame to Store Result ----------- #
duplicateSummary = pd.DataFrame({'Count': [duplicatesCount]}, index = ['Duplicated Rows'])

# ----------- Function for Colour Coding ----------- #
def colour_code(val):
    if val > 0:
        return 'background-color: red; color: white'
    else:
        return 'background-color: green; color: white'

# ----------- Apply Colour Coding to the 'Count' Column ----------- #
duplicateSummary = duplicateSummary.style.map(colour_code, subset=['Count'])

# ----------- Display the Result ----------- #
duplicateSummary

,Count
Duplicated Rows,0


<p>From the output of 0 as seen above, I am able to verify that all duplicated data has been successfully removed and will no logner affect any data analysis conducted in the later part.</p>

<hr/>
<h2>2.6 Fixing Data Types</h2>

<p>Data Types differences within the dataframe needs to be addressed and adjusted on a case by case basis depending on the context of the data column.</p>

<hr/>
<h3>2.6.1 Identifying Potential Incorrect Data Types</h3>

<p>Before being able to fix the data types, I will have to indentify the data columns with the wrong data type. This will allow me to subsequently fix the data type for each column.</p>

In [118]:
# ----------- Display Data Types ----------- #
merged_df.dtypes

company_name                   object
title                          object
max_salary                    float64
pay_period                     object
location                       object
views                         float64
min_salary                    float64
formatted_work_type            object
applies                       float64
remote_allowed                float64
application_type               object
expiry                        float64
formatted_experience_level     object
listed_time                   float64
dtype: object

<p>From the output above, I am able to identify some data columns that needs to have their respective data types adjusted due to either the current data type being unsuitable or incorrect. The data columns that needs to be adjusted are listed below.</p>
<br/>
<p><strong><u>max_salary</u></strong></p>
<ul>
    <li>Currrent: float64</li>
    <li>Intended: int64</li>
    <li>Reason: Comprises of mainly integers with no decimal points.</li>
</ul>
<p><strong><u>views</u></strong></p>
<ul>
    <li>Currrent: float64</li>
    <li>Intended: int64</li>
    <li>Reason: Comprises of whole numbers.</li>
</ul>
<p><strong><u>min_salary</u></strong></p>
<ul>
    <li>Currrent: float64</li>
    <li>Intended: int64</li>
    <li>Reason: Comprises of mainly integers with no decimal points.</li>
</ul>
<p><strong><u>applies</u></strong></p>
<ul>
    <li>Currrent: float64</li>
    <li>Intended: int64</li>
    <li>Reason: Comprises of whole numbers.</li>
</ul>
<p><strong><u>remote_allowed</u></strong></p>
<ul>
    <li>Currrent: float64</li>
    <li>Intended: int64</li>
    <li>Reason: It is a dummy variable, there should only be either 1 or 0.</li>
</ul>
<p><strong><u>expiry</u></strong></p>
<ul>
    <li>Currrent: float64</li>
    <li>Intended: datetime64</li>
    <li>Reason: The data column represents timestamps.</li>
</ul>
<p><strong><u>listed_time</u></strong></p>
<ul>
    <li>Currrent: float64</li>
    <li>Intended: datetime64</li>
    <li>Reason: The data column represents timestamps.</li>
</ul>
<br/>
<p>With the incorrect and unsuitable data types identified, I will now proceed with rectifying each of the data types.</p>

<hr/>
<h3>2.6.2 Addressing Incorrect and Unsuitable Data Types</h3>

<p>In the previous section, I have identified 7 data columns that needs to have its data type rectified. The operation will be done within the cell below.</p>

In [122]:
# ----------- Rectifying Numerical Columns ----------- #
merged_df['max_salary'] = merged_df['max_salary'].astype('int64')
merged_df['views'] = merged_df['views'].astype('int64')
merged_df['min_salary'] = merged_df['min_salary'].astype('int64')
merged_df['applies'] = merged_df['applies'].astype('int64')
merged_df['remote_allowed'] = merged_df['remote_allowed'].astype('int64')

# ----------- Converting Timestamps to Datetime ----------- #
merged_df['expiry'] = pd.to_datetime(merged_df['expiry'], unit='ms', errors='coerce')
merged_df['listed_time'] = pd.to_datetime(merged_df['listed_time'], unit='ms', errors='coerce')

# ----------- Data Verification ----------- #
merged_df.dtypes

company_name                          object
title                                 object
max_salary                             int64
pay_period                            object
location                              object
views                                  int64
min_salary                             int64
formatted_work_type                   object
applies                                int64
remote_allowed                         int64
application_type                      object
expiry                        datetime64[ns]
formatted_experience_level            object
listed_time                   datetime64[ns]
dtype: object

<p>From the output above, I am able to verify that all data types that needs to be rectified have been rectified accordingly. Therefore, I will move on to Standardising Categorical Columns.</p>

<hr/>
<h2>2.7 Standardising Categorical Data</h2>

<p>To ensure a consistent data format throughout each data columns, I will be changing some data columns to make it visually appealing and consistent.</p>

<hr/>
<h3>2.7.1 Identifying Types of Pay Period</h3>

<p>Before being able to standardise and format the 'pay_period' data column, I will need to identify what data are within the data column. In the cell below I will be finding out the different pay period and subsequently format them.</p>

In [128]:
# ----------- Finding The Various Quantity ----------- #
pay_period_counts = merged_df['pay_period'].value_counts().reset_index()

# ----------- Renaming of Data Columns ----------- #
pay_period_counts.columns = ['pay_period', 'count']

# ----------- Setting New Index ----------- #
pay_period_counts = pay_period_counts.set_index('pay_period')

# ----------- Displaying of Results ----------- #
pay_period_counts

,count
pay_period,
YEARLY,108373
HOURLY,14703
MONTHLY,516
WEEKLY,171
BIWEEKLY,9


<p>With reference to the output of the dataframe above, I am able to see that there is a total of 5 different pay periods which I have to standardise and format.</p>

<hr/>
<h3>2.7.2 Standardisation of Pay Period Format</h3>

<p>Upon verifying the content of the 'pay_period' data column's content, I will proceed to standardise the format of each of the data in the cell below according to their original data value. In the list below, I will indicate how I plan to format them.</p>
<br/>
<p>YEARLY: <strong><u>Yearly</u></strong></p>
<p>MONTHLY: <strong><u>Monthly</u></strong></p>
<p>BIWEEKLY: <strong><u>Biweekly</u></strong></p>
<p>WEEKLY: <strong><u>Weekly</u></strong></p>
<p>HOURLY: <strong><u>Hourly</u></strong></p>
<br/>
<p>With the new format listed, I will proceed to perform the replace oepration in the cell below.</p>

In [299]:
# ----------- Standardising Text Format ----------- #
merged_df['pay_period'] = merged_df['pay_period'].replace({'YEARLY': 'Yearly', 'MONTHLY': 'Monthly', 'BIWEEKLY': 'Biweekly',
                                                           'WEEKLY': 'Weekly', 'HOURLY': 'Hourly'})

# ----------- Verification of Changes ----------- #
pay_period_counts = merged_df['pay_period'].value_counts().reset_index()

# ----------- Renaming of Data Columns ----------- #
pay_period_counts.columns = ['pay_period', 'count']

# ----------- Setting New Index ----------- #
pay_period_counts = pay_period_counts.set_index('pay_period')

# ----------- Displaying of Results ----------- #
pay_period_counts

,count
pay_period,
Yearly,97690
Hourly,14348
Monthly,485
Weekly,170
Biweekly,3


<p>From the dataframe above, I am able to see that the changes have been made successfully.</p>

<hr/>
<h2>2.8 Feature Engineering</h2>

<p>In Feature Engineering, I will be creating a few new features that will make the Data Analysis more steamlined and interpretable. This is because, new features can help in cutting down the amount of data visualisations through potentially combining 2 data columns into 1 single new data columns. The new features that I will be creating are listed below. Please note I will be making the assumption that each year has 52 Weeks and each worker works 40 Hours a week.</p>

$$
\text{max\_salary\_normalised} = \text{max\_salary (Monthly Pay)} \times 12
$$
<br/>
$$
\text{max\_salary\_normalised} = \text{max\_salary (Biweekly Pay)} \times 26
$$
<br/>
$$
\text{max\_salary\_normalised} = \text{max\_salary (Weekly Pay)} \times 52
$$
<br/>
$$
\text{max\_salary\_normalised} = \text{max\_salary (Hourly Pay)} \times 40 \times 52
$$
<br/>
$$
\text{min\_salary\_normalised} = \text{min\_salary (Monthly Pay)} \times 12
$$
<br/>
$$
\text{min\_salary\_normalised} = \text{min\_salary (Biweekly Pay)} \times 26
$$
<br/>
$$
\text{min\_salary\_normalised} = \text{min\_salary (Weekly Pay)} \times 52
$$
<br/>
$$
\text{min\_salary\_normalised} = \text{min\_salary (Hourly Pay)} \times 40 \times 52
$$
<br/>
$$
\text{postingDuration} = \text{expiry} - \text{listed\_time}
$$
<br/>
$$
\text{salaryRange} = \text{max\_salary\_normalised} - \text{min\_salary\_normalised}
$$
<br/>
$$
\text{applyViewRatio} = \frac{\text{applies}}{\text{views}}
$$
<br/>
<p>With the formulas for each new feature listed out, I will proceed with creating the new features in the cell below.</p>

In [136]:
# ----------- Function to Normalise Salary With Respect to Pay Period ----------- #
def normalise_salary_max(row):
    if row['pay_period'] == 'Hourly':
        return row['max_salary'] * 40 * 52  # 40 Hours/Week, 52 Weeks/Year
    elif row['pay_period'] == 'Monthly':
        return row['max_salary'] * 12       # 12 Months/Year
    elif row['pay_period'] == 'Biweekly':
        return row['max_salary'] * 26       # 26 Weeks/Year
    elif row['pay_period'] == 'Weekly':
        return row['max_salary'] * 52       # 52 Weeks/Year
    else:
        return row['max_salary']            # No Conversion Needed

# ----------- Creation of New Data Column 'max_salary_normalised' ----------- #
merged_df['max_salary_normalised'] = merged_df.apply(normalise_salary_max, axis=1)

# ----------- Function to Normalise Salary With Respect to Pay Period ----------- #
def normalise_salary_min(row):
    if row['pay_period'] == 'Hourly':
        return row['min_salary'] * 40 * 52  # 40 hours/week, 52 weeks/year
    elif row['pay_period'] == 'Monthly':
        return row['min_salary'] * 12       # 12 months/year
    elif row['pay_period'] == 'Biweekly':
        return row['min_salary'] * 26       # 26 Weeks/Year
    elif row['pay_period'] == 'Weekly':
        return row['min_salary'] * 52       # 52 Weeks/Year
    else:
        return row['min_salary']            # No conversion needed

# ----------- Creation of New Data Column 'min_salary_normalised' ----------- #
merged_df['min_salary_normalised'] = merged_df.apply(normalise_salary_min, axis = 1)

# ----------- Posting Duration as Alias postingDuration ----------- #
merged_df['postingDuration'] = (merged_df['expiry'] - merged_df['listed_time']).dt.days

# ----------- Salary Range as Alias salaryRange ----------- #
merged_df['salary_range'] = merged_df['max_salary_normalised'] - merged_df['min_salary_normalised']

# ----------- Apply to View Ratio as Alias applyViewRation ----------- #
merged_df['apply_view_ratio'] = merged_df['applies'] / merged_df['views']
merged_df['apply_view_ratio'] = merged_df['applies'] / merged_df['views'].replace(0, np.nan)
merged_df['apply_view_ratio'] = merged_df['apply_view_ratio'].fillna(0)  # Replace NaN with 0 after calculation


# ----------- Displaying Dataframe ----------- #
merged_df.head(5)

,company_name,title,max_salary,pay_period,location,views,min_salary,formatted_work_type,applies,remote_allowed,application_type,expiry,formatted_experience_level,listed_time,max_salary_normalised,min_salary_normalised,postingDuration,salary_range,apply_view_ratio
job_id,,,,,,,,,,,,,,,,,,,
3904387609,Sherwin-Williams,"Store Sales Representative - Chesterfield, VA",66976,Biweekly,"Chesterfield, VA",7,54009,Full-time,0,0,OffsiteApply,2024-05-17 23:53:20,Entry level,2024-04-17 21:40:00,1741376,1404234,30,337142,0.000000
3904388204,Sherwin-Williams,R&D Scientific Programmer,89965,Biweekly,"Cleveland, OH",31,71323,Full-time,4,1,OffsiteApply,2024-05-17 23:53:20,Entry level,2024-04-17 21:40:00,2339090,1854398,30,484692,0.129032
3904991870,Sherwin-Williams,"Sales Representative, Residential Repaint",87660,Biweekly,"Waltham, MA",4,69496,Full-time,0,0,OffsiteApply,2024-05-18 16:33:20,Entry level,2024-04-18 17:06:40,2279160,1806896,29,472264,0.000000
3904995353,Sherwin-Williams,"Sales Representative, Residential Repaint",3366,Biweekly,"Hammonton, NJ",4,2245,Full-time,0,0,OffsiteApply,2024-05-18 16:33:20,Entry level,2024-04-18 17:06:40,87516,58370,29,29146,0.000000
3904996213,Sherwin-Williams,"Sales Representative, Residential Repaint",81434,Biweekly,"Dedham, MA",4,64519,Full-time,0,0,OffsiteApply,2024-05-18 16:33:20,Entry level,2024-04-18 17:06:40,2117284,1677494,29,439790,0.000000


<p>With reference to the output of the dataframe above, I am able to verify that the 4 new features have been successfulyl created and I am able to move on to Data Analysis.</p>

<hr/>
<h2>2.9 Handling Outliers</h2>

<p>In this sub-section, I will be handling and subsequently removing outliers that are within the dataframe. Outliers can make Data Analysis skewed and potentially affect the accuracy and reliability of the results. Therefore, it is of the utmost importance that Outliers be addressed and subsequently removed if needed. As there are multiple ways to address outliers, I will be comparing 2 ways of removal which are using the Upper / Lower Fence Method (IQR Method) and the Z-Score method. After comparing, I will evaluate which method will be most reliable and subsequently use that method.</p>

<hr/>
<h3>2.9.1 Handling Outliers Using IQR Method</h3>

<p>The IQR Method requires the establishment of a Upper Fence and Lower Fence such that any data which are above the Upper Fence will be deemed invalid and illegitimate and will be removed whereas any data which are below the Lower Fence will be deemed likewise and removed. In the cells below, I will be removing Outliers using the IQR Method and subsequently comparing the effectiveness and results with the Z-Score Method. The formula for establishing the Upper Fence, Lower Fence and IQR are listed below.</p>

$$\text{IQR} = \text{Q3} - \text{Q1}$$
$$\text{Lower Fence} = \text{Q1} - 1.5 \times \text{IQR}$$
$$\text{Upper Fence} = \text{Q3} + 1.5 \times \text{IQR}$$
<br/>
<p>With the formulas listed, I will proceed with identifying and removing the Outliers.</p>

<hr/>
<h4>2.9.1.1 Identifying Outliers Using IQR Method For Maximum Salary</h4>

<p>Before being able to address and remove Outliers, I will firstly have to indentify the outliers for the relevant data columns. As Outliers can only be identified and addressed for Numerical Data, I will only be identifying the Outliers for 'min_salary', 'max_salary' and 'salary_range' as there are no true limit for views and applications. However, as salary is paid in different periods, I will be using data column called 'max_salary_normalised'. In using this new data column, a few assumptions will be used.</p>
<br/>
<p><strong><u>Assumptions</u></strong></p>
<ul>
    <li>There is a total of 40 hours of work done per week.</li>
    <li>There is a total of 52 weeks in a year.</li>
</ul>
<br/>
<p>With the assumptions in place, I will proceed to make the new data columns and identify the outliers within the data column.</p>

In [144]:
# ----------- Identifying Outliers for min_salary_normalised using IQR ----------- #
Q1 = merged_df['min_salary_normalised'].quantile(0.25)
Q3 = merged_df['min_salary_normalised'].quantile(0.75)
IQR = Q3 - Q1

# ----------- Defining Upper Fence and Lower Fence ----------- #
lowerFence = Q1 - 1.5 * IQR
upperFence = Q3 + 1.5 * IQR

# ----------- Extracting Outliers ----------- #
min_salary_outliers_IQR = merged_df[(merged_df['min_salary_normalised'] < lowerFence) |
                                    (merged_df['min_salary_normalised'] > upperFence)]

# ----------- Displaying Outliers ----------- #
print(f"Number of Outliers in min_salary_normalised using IQR: {len(min_salary_outliers_IQR)}")
min_salary_outliers_IQR.head(5)

Number of Outliers in min_salary_normalised using IQR: 3803


,company_name,title,max_salary,pay_period,location,views,min_salary,formatted_work_type,applies,remote_allowed,application_type,expiry,formatted_experience_level,listed_time,max_salary_normalised,min_salary_normalised,postingDuration,salary_range,apply_view_ratio
job_id,,,,,,,,,,,,,,,,,,,
3904387609,Sherwin-Williams,"Store Sales Representative - Chesterfield, VA",66976,Biweekly,"Chesterfield, VA",7,54009,Full-time,0,0,OffsiteApply,2024-05-17 23:53:20,Entry level,2024-04-17 21:40:00,1741376,1404234,30,337142,0.000000
3904388204,Sherwin-Williams,R&D Scientific Programmer,89965,Biweekly,"Cleveland, OH",31,71323,Full-time,4,1,OffsiteApply,2024-05-17 23:53:20,Entry level,2024-04-17 21:40:00,2339090,1854398,30,484692,0.129032
3904991870,Sherwin-Williams,"Sales Representative, Residential Repaint",87660,Biweekly,"Waltham, MA",4,69496,Full-time,0,0,OffsiteApply,2024-05-18 16:33:20,Entry level,2024-04-18 17:06:40,2279160,1806896,29,472264,0.000000
3904996213,Sherwin-Williams,"Sales Representative, Residential Repaint",81434,Biweekly,"Dedham, MA",4,64519,Full-time,0,0,OffsiteApply,2024-05-18 16:33:20,Entry level,2024-04-18 17:06:40,2117284,1677494,29,439790,0.000000
3905333190,Sherwin-Williams,"Store Sales Representative - Cleveland, OH",66976,Biweekly,"Cleveland, OH",4,54009,Full-time,0,0,OffsiteApply,2024-05-18 22:06:40,Entry level,2024-04-18 22:40:00,1741376,1404234,29,337142,0.000000


<p>With reference to the output above, I am able to see that there are a total of 33825 Outliers within the dataframe. I will move on to check for outliers for min_salary and subsequently remove these outliers</p>

<hr/>
<h4>2.9.1.2 Identifying Outliers Using IQR Method For Minimum Salary</h4>

<p>Before being able to address and remove Outliers, I will firstly have to indentify the outliers for the relevant data columns. As Outliers can only be identified and addressed for Numerical Data, I will only be identifying the Outliers for 'min_salary', 'max_salary' and 'salary_range' as there are no true limit for views and applications. However, as salary is paid in different periods, I will be using data column called 'min_salary_normalised'. In using this new data column, a few assumptions will be used.</p>
<br/>
<p><strong><u>Assumptions</u></strong></p>
<ul>
    <li>There is a total of 40 hours of work done per week.</li>
    <li>There is a total of 52 weeks in a year.</li>
</ul>
<br/>
<p>With the assumptions in place, I will proceed to make the new data columns and identify the outliers within the data column.</p>

In [148]:
# ----------- Identifying Outliers for min_salary_normalised using IQR ----------- #
Q1 = merged_df['max_salary_normalised'].quantile(0.25)
Q3 = merged_df['max_salary_normalised'].quantile(0.75)
IQR = Q3 - Q1

# ----------- Defining Upper Fence and Lower Fence ----------- #
lowerFence = Q1 - 1.5 * IQR
upperFence = Q3 + 1.5 * IQR

# ----------- Extracting Outliers ----------- #
max_salary_outliers_IQR = merged_df[(merged_df['max_salary_normalised'] < lowerFence) |
                                    (merged_df['max_salary_normalised'] > upperFence)]

# ----------- Displaying Outliers ----------- #
print(f"Number of Outliers in min_salary_normalised using IQR: {len(max_salary_outliers_IQR)}")
max_salary_outliers_IQR.head(5)

Number of Outliers in min_salary_normalised using IQR: 4327


,company_name,title,max_salary,pay_period,location,views,min_salary,formatted_work_type,applies,remote_allowed,application_type,expiry,formatted_experience_level,listed_time,max_salary_normalised,min_salary_normalised,postingDuration,salary_range,apply_view_ratio
job_id,,,,,,,,,,,,,,,,,,,
3904387609,Sherwin-Williams,"Store Sales Representative - Chesterfield, VA",66976,Biweekly,"Chesterfield, VA",7,54009,Full-time,0,0,OffsiteApply,2024-05-17 23:53:20,Entry level,2024-04-17 21:40:00,1741376,1404234,30,337142,0.000000
3904388204,Sherwin-Williams,R&D Scientific Programmer,89965,Biweekly,"Cleveland, OH",31,71323,Full-time,4,1,OffsiteApply,2024-05-17 23:53:20,Entry level,2024-04-17 21:40:00,2339090,1854398,30,484692,0.129032
3904991870,Sherwin-Williams,"Sales Representative, Residential Repaint",87660,Biweekly,"Waltham, MA",4,69496,Full-time,0,0,OffsiteApply,2024-05-18 16:33:20,Entry level,2024-04-18 17:06:40,2279160,1806896,29,472264,0.000000
3904996213,Sherwin-Williams,"Sales Representative, Residential Repaint",81434,Biweekly,"Dedham, MA",4,64519,Full-time,0,0,OffsiteApply,2024-05-18 16:33:20,Entry level,2024-04-18 17:06:40,2117284,1677494,29,439790,0.000000
3905333190,Sherwin-Williams,"Store Sales Representative - Cleveland, OH",66976,Biweekly,"Cleveland, OH",4,54009,Full-time,0,0,OffsiteApply,2024-05-18 22:06:40,Entry level,2024-04-18 22:40:00,1741376,1404234,29,337142,0.000000


<p>With reference to the output above, I am able to see that there are a total of 33825 Outliers within the dataframe. I will move on to check for outliers for min_salary and subsequently remove these outliers</p>

<hr/>
<h4>2.9.1.3 Identifying Outliers Using IQR Method For Salary Range</h4>

<p>Before being able to address and remove Outliers, I will firstly have to indentify the outliers for the relevant data columns. As Outliers can only be identified and addressed for Numerical Data, I will only be identifying the Outliers for 'min_salary', 'max_salary' and 'salary_range' as there are no true limit for views and applications. However, as salary is paid in different periods, I will be using data column called 'salary_range'. In using this new data column, a few assumptions will be used as normalised salary have been used to calculate salary range.</p>
<br/>
<p><strong><u>Assumptions</u></strong></p>
<ul>
    <li>There is a total of 40 hours of work done per week.</li>
    <li>There is a total of 52 weeks in a year.</li>
</ul>
<br/>
<p>With the assumptions in place, I will proceed to make the new data columns and identify the outliers within the data column.</p>

In [152]:
# ----------- Identifying Outliers for min_salary_normalised using IQR ----------- #
Q1 = merged_df['salary_range'].quantile(0.25)
Q3 = merged_df['salary_range'].quantile(0.75)
IQR = Q3 - Q1

# ----------- Defining Upper Fence and Lower Fence ----------- #
lowerFence = Q1 - 1.5 * IQR
upperFence = Q3 + 1.5 * IQR

# ----------- Extracting Outliers ----------- #
salary_range_outliers_IQR = merged_df[(merged_df['salary_range'] < lowerFence) |
                                    (merged_df['salary_range'] > upperFence)]

# ----------- Displaying Outliers ----------- #
print(f"Number of Outliers in min_salary_normalised using IQR: {len(salary_range_outliers_IQR)}")
salary_range_outliers_IQR.head(5)

Number of Outliers in min_salary_normalised using IQR: 7090


,company_name,title,max_salary,pay_period,location,views,min_salary,formatted_work_type,applies,remote_allowed,application_type,expiry,formatted_experience_level,listed_time,max_salary_normalised,min_salary_normalised,postingDuration,salary_range,apply_view_ratio
job_id,,,,,,,,,,,,,,,,,,,
3904387609,Sherwin-Williams,"Store Sales Representative - Chesterfield, VA",66976,Biweekly,"Chesterfield, VA",7,54009,Full-time,0,0,OffsiteApply,2024-05-17 23:53:20,Entry level,2024-04-17 21:40:00,1741376,1404234,30,337142,0.000000
3904388204,Sherwin-Williams,R&D Scientific Programmer,89965,Biweekly,"Cleveland, OH",31,71323,Full-time,4,1,OffsiteApply,2024-05-17 23:53:20,Entry level,2024-04-17 21:40:00,2339090,1854398,30,484692,0.129032
3904991870,Sherwin-Williams,"Sales Representative, Residential Repaint",87660,Biweekly,"Waltham, MA",4,69496,Full-time,0,0,OffsiteApply,2024-05-18 16:33:20,Entry level,2024-04-18 17:06:40,2279160,1806896,29,472264,0.000000
3904996213,Sherwin-Williams,"Sales Representative, Residential Repaint",81434,Biweekly,"Dedham, MA",4,64519,Full-time,0,0,OffsiteApply,2024-05-18 16:33:20,Entry level,2024-04-18 17:06:40,2117284,1677494,29,439790,0.000000
3905333190,Sherwin-Williams,"Store Sales Representative - Cleveland, OH",66976,Biweekly,"Cleveland, OH",4,54009,Full-time,0,0,OffsiteApply,2024-05-18 22:06:40,Entry level,2024-04-18 22:40:00,1741376,1404234,29,337142,0.000000


<p>With reference to the datafram above, I am able to verify that all the Outliers have been successfully removed and addressed.</p>

<hr/>
<h4>2.9.1.4 Removing Identified Outliers Using IQR Method</h4>

<p>After identifying large amounts of Outliers, I will now proceed to remove the Outliers from the dataframe. However, I will be doing this operation on a copy of the original dataframe to ensure I am still able to do the Outliers Identification and Removal using the Z-Score method.</p>

In [156]:
# ----------- Copy the Main DataFrame ----------- #
IQRdf = merged_df.copy()

# ----------- Define a Function to Remove Outliers Using IQR ----------- #
def remove_outliers_iqr(df, column):
    # Calculate Q1, Q3, and IQR
    Q1 = df[column].quantile(0.25)
    Q3 = df[column].quantile(0.75)
    IQR = Q3 - Q1
    
    # Define lower and upper bounds
    lowerFence = Q1 - 1.5 * IQR
    upperFence = Q3 + 1.5 * IQR
    
    # Filter DataFrame to keep only non-outliers
    df_no_outliers = df[(df[column] >= lowerFence) & (df[column] <= upperFence)]
    
    print(f"Outliers removed for '{column}': {len(df) - len(df_no_outliers)} rows")
    return df_no_outliers

# ----------- Remove Outliers for Each Column ----------- #
IQRdf = remove_outliers_iqr(IQRdf, 'min_salary_normalised')
IQRdf = remove_outliers_iqr(IQRdf, 'max_salary_normalised')
IQRdf = remove_outliers_iqr(IQRdf, 'salary_range')

# ----------- Display Results ----------- #
print(f"Rows after outlier removal: {len(IQRdf)}")
IQRdf.head(5)

Outliers removed for 'min_salary_normalised': 3803 rows
Outliers removed for 'max_salary_normalised': 2239 rows
Outliers removed for 'salary_range': 5034 rows
Rows after outlier removal: 112696


,company_name,title,max_salary,pay_period,location,views,min_salary,formatted_work_type,applies,remote_allowed,application_type,expiry,formatted_experience_level,listed_time,max_salary_normalised,min_salary_normalised,postingDuration,salary_range,apply_view_ratio
job_id,,,,,,,,,,,,,,,,,,,
3904995353,Sherwin-Williams,"Sales Representative, Residential Repaint",3366,Biweekly,"Hammonton, NJ",4,2245,Full-time,0,0,OffsiteApply,2024-05-18 16:33:20,Entry level,2024-04-18 17:06:40,87516,58370,29,29146,0.000000
3895598632,Sherwin-Williams,Assistant Manager Trainee,2057,Biweekly,"Sioux Falls, SD",7,2057,Full-time,0,0,OffsiteApply,2024-05-12 05:00:00,Internship,2024-04-12 05:33:20,53482,53482,29,0,0.000000
3904386637,Sherwin-Williams,Assistant Manager Trainee,2134,Biweekly,"Orange, CA",7,2134,Full-time,0,0,OffsiteApply,2024-05-17 23:53:20,Internship,2024-04-17 21:40:00,55484,55484,30,0,0.000000
3863624365,Conquer Employment,Technical Planner,28,Hourly,"Terrell, TX",15,24,Full-time,0,0,ComplexOnsiteApply,2024-05-05 20:13:20,Associate,2024-04-05 20:46:40,58240,49920,29,8320,0.000000
3884430611,Planet Interactive,Digital Designer,40,Hourly,"Bolingbrook, IL",29,35,Contract,4,0,ComplexOnsiteApply,2024-05-05 20:13:20,Associate,2024-04-05 20:46:40,83200,72800,29,10400,0.137931


<p>With reference to the output of the dataframe above, I am able to verify that the Outliers have been successfully removed. I will now move on to the next method which is the Z-Score method.</p>

<hr/>
<h3>2.9.2 Handling Outliers Using Z-Score</h3>

<p>In the previous section, I have used the IQR method to identify and address Outliers. However, as there are many methods to identify and address Outliers, I will be using the Z-Score method in this section. After removing Outliers using the Z-score Method, I will then cross refer the result and verify which method is most suitable. The formula for Z-Score is indicated below.</p>

$$
Z = \frac{X - \mu}{\sigma}
$$

\begin{align*}
Z & : \text{Z-Score (standardised value)} \\
X & : \text{Observed data point} \\
\mu & : \text{Mean of the data} \\
\sigma & : \text{Standard deviation of the data}
\end{align*}

<br>
<p>With the formula indicated, I will proceed to identify the Outliers and remove the Outliers in the following cells.</p>

<hr/>
<h4>2.9.2.1 Identifying Outliers Using Z-Score For Maximum Salary</h4>

<p>Before being able to address and remove Outliers, I will firstly have to indentify the outliers for the relevant data columns. As Outliers can only be identified and addressed for Numerical Data, I will only be identifying the Outliers for 'min_salary', 'max_salary' and 'salary_range' as there are no true limit for views and applications. However, as salary is paid in different periods, I will be using data column called 'max_salary_normalised'.  I will also be utilising the threshold of $\pm$ 3 to identify the Outliers. In using this new data column, a few assumptions will be used.</p>
<br/>
<p><strong><u>Assumptions</u></strong></p>
<ul>
    <li>There is a total of 40 hours of work done per week.</li>
    <li>There is a total of 52 weeks in a year.</li>
    <li>Empirical Rule applies.</li>
</ul>
<br/>
<p>With the assumptions in place, I will proceed to make the new data columns and identify the outliers within the data column.</p>

In [162]:
# ----------- Calculating Z-Score's Variables ----------- #
mean_max = merged_df['max_salary_normalised'].mean()
std_max = merged_df['max_salary_normalised'].std()

# ----------- Calculating Z-Score ----------- #
merged_df['z_score_max_salary'] = (merged_df['max_salary_normalised'] - mean_max) / std_max

# ----------- Identifying Outliers ----------- #
max_salary_outliers_z = merged_df[merged_df['z_score_max_salary'].abs() > 3]

# ----------- Dropping Data Column After Usage ----------- #
merged_df = merged_df.drop(columns=['z_score_max_salary'])

# ----------- Displaying Results ----------- #
print(f"Number of Outliers in max_salary_normalised: {len(max_salary_outliers_z)}")
max_salary_outliers_z

Number of Outliers in max_salary_normalised: 76


,company_name,title,max_salary,pay_period,location,views,min_salary,formatted_work_type,applies,remote_allowed,application_type,expiry,formatted_experience_level,listed_time,max_salary_normalised,min_salary_normalised,postingDuration,salary_range,apply_view_ratio,z_score_max_salary
job_id,,,,,,,,,,,,,,,,,,,,
3886471794,Eastridge Workforce Solutions,Intellectual Property Associate (246215),275000,Hourly,"Dallas, TX",4,240000,Full-time,0,0,SimpleOnsiteApply,2024-05-09 04:46:40,Entry level,2024-04-09 05:20:00,572000000,499200000,29,72800000,0.000000,155.893454
3889115413,Nike,"Store Associate, Seas - Foxwoods",12518,Hourly,"Mashantucket, CT",5,11268,Full-time,0,0,OffsiteApply,2024-05-09 04:46:40,Entry level,2024-04-09 05:20:00,26037440,23437440,29,2600000,0.000000,7.046088
3889116357,Nike,"Store Associate, FT - Vineland",25012,Hourly,"Orlando, FL",7,22512,Full-time,0,0,OffsiteApply,2024-05-09 04:46:40,Entry level,2024-04-09 05:20:00,52024960,46824960,29,5200000,0.000000,14.131141
3889117075,Nike,"Store Associate, PT - Finger Lakes",37506,Hourly,"Waterloo, NY",6,33756,Part-time,0,0,OffsiteApply,2024-05-09 04:46:40,Entry level,2024-04-09 05:20:00,78012480,70212480,29,7800000,0.000000,21.216195
3889117757,The Hillman Group,Full Time Field Service Technician – Wilmingto...,50000,Hourly,"Wilmington, NC",8,45000,Full-time,0,0,OffsiteApply,2024-05-09 04:46:40,Entry level,2024-04-09 05:20:00,104000000,93600000,29,10400000,0.000000,28.301248
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
3902879426,Adams & Martin Group,Trial Paralegal - Hybrid in OC or LA!,100000,Hourly,"Los Angeles County, CA",2,80000,Full-time,0,0,SimpleOnsiteApply,2024-05-19 23:06:40,Mid-Senior level,2024-04-19 23:40:00,208000000,166400000,29,41600000,0.000000,56.655072
3904906761,Unknown,Content Writer,40046666,Yearly,"Virginia, United States",70,28373333,Internship,20,1,SimpleOnsiteApply,2024-05-18 13:46:40,Mid-Senior level,2024-04-18 14:20:00,40046666,28373333,29,11673333,0.285714,10.865464
3904911266,Independent Sports & Entertainment (ISE),Client Relations Manager,80023333,Yearly,"Chicago, IL",22,56686666,Full-time,5,0,ComplexOnsiteApply,2024-10-15 13:46:40,Mid-Senior level,2024-04-18 14:20:00,80023333,56686666,179,23336667,0.227273,21.764420


<p>With reference to the output of the dataframe, I am able to see that the Outlier has been identified successfully.</p>

<hr/>
<h4>2.9.2.2 Identifying Outliers Using Z-Score For Minimum Salary</h4>

<p>Before being able to address and remove Outliers, I will firstly have to indentify the outliers for the relevant data columns. As Outliers can only be identified and addressed for Numerical Data, I will only be identifying the Outliers for 'min_salary', 'max_salary' and 'salary_range' as there are no true limit for views and applications. However, as salary is paid in different periods, I will be using data column called 'min_salary_normalised'.  I will also be utilising the threshold of $\pm$ 3 to identify the Outliers. In using this new data column, a few assumptions will be used.</p>
<br/>
<p><strong><u>Assumptions</u></strong></p>
<ul>
    <li>There is a total of 40 hours of work done per week.</li>
    <li>There is a total of 52 weeks in a year.</li>
    <li>Empirical Rule applies.</li>
</ul>
<br/>
<p>With the assumptions in place, I will proceed to make the new data columns and identify the outliers within the data column.</p>

In [166]:
# ----------- Calculating Z-Score's Variables ----------- #
mean_max = merged_df['min_salary_normalised'].mean()
std_max = merged_df['min_salary_normalised'].std()

# ----------- Calculating Z-Score ----------- #
merged_df['z_score_min_salary'] = (merged_df['min_salary_normalised'] - mean_max) / std_max

# ----------- Identifying Outliers ----------- #
min_salary_outliers_z = merged_df[merged_df['z_score_min_salary'].abs() > 3]

# ----------- Dropping Data Column After Usage ----------- #
merged_df = merged_df.drop(columns=['z_score_min_salary'])

# ----------- Displaying Results ----------- #
print(f"Number of Outliers in min_salary_normalised: {len(min_salary_outliers_z)}")
min_salary_outliers_z

Number of Outliers in min_salary_normalised: 75


,company_name,title,max_salary,pay_period,location,views,min_salary,formatted_work_type,applies,remote_allowed,application_type,expiry,formatted_experience_level,listed_time,max_salary_normalised,min_salary_normalised,postingDuration,salary_range,apply_view_ratio,z_score_min_salary
job_id,,,,,,,,,,,,,,,,,,,,
3886471794,Eastridge Workforce Solutions,Intellectual Property Associate (246215),275000,Hourly,"Dallas, TX",4,240000,Full-time,0,0,SimpleOnsiteApply,2024-05-09 04:46:40,Entry level,2024-04-09 05:20:00,572000000,499200000,29,72800000,0.000000,161.955512
3889115413,Nike,"Store Associate, Seas - Foxwoods",12518,Hourly,"Mashantucket, CT",5,11268,Full-time,0,0,OffsiteApply,2024-05-09 04:46:40,Entry level,2024-04-09 05:20:00,26037440,23437440,29,2600000,0.000000,7.558961
3889116357,Nike,"Store Associate, FT - Vineland",25012,Hourly,"Orlando, FL",7,22512,Full-time,0,0,OffsiteApply,2024-05-09 04:46:40,Entry level,2024-04-09 05:20:00,52024960,46824960,29,5200000,0.000000,15.148782
3889117075,Nike,"Store Associate, PT - Finger Lakes",37506,Hourly,"Waterloo, NY",6,33756,Part-time,0,0,OffsiteApply,2024-05-09 04:46:40,Entry level,2024-04-09 05:20:00,78012480,70212480,29,7800000,0.000000,22.738602
3889117757,The Hillman Group,Full Time Field Service Technician – Wilmingto...,50000,Hourly,"Wilmington, NC",8,45000,Full-time,0,0,OffsiteApply,2024-05-09 04:46:40,Entry level,2024-04-09 05:20:00,104000000,93600000,29,10400000,0.000000,30.328423
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
3902879426,Adams & Martin Group,Trial Paralegal - Hybrid in OC or LA!,100000,Hourly,"Los Angeles County, CA",2,80000,Full-time,0,0,SimpleOnsiteApply,2024-05-19 23:06:40,Mid-Senior level,2024-04-19 23:40:00,208000000,166400000,29,41600000,0.000000,53.953798
3904906761,Unknown,Content Writer,40046666,Yearly,"Virginia, United States",70,28373333,Internship,20,1,SimpleOnsiteApply,2024-05-18 13:46:40,Mid-Senior level,2024-04-18 14:20:00,40046666,28373333,29,11673333,0.285714,9.160779
3904911266,Independent Sports & Entertainment (ISE),Client Relations Manager,80023333,Yearly,"Chicago, IL",22,56686666,Full-time,5,0,ComplexOnsiteApply,2024-10-15 13:46:40,Mid-Senior level,2024-04-18 14:20:00,80023333,56686666,179,23336667,0.227273,18.349146


<p>With reference to the output of the dataframe, I am able to see that the Outlier has been identified successfully.</p>

<h4>3.9.2.3 Identifying Outliers Using Z-Score For Salary Range</h4>

<p>Before being able to address and remove Outliers, I will firstly have to indentify the outliers for the relevant data columns. As Outliers can only be identified and addressed for Numerical Data, I will only be identifying the Outliers for 'min_salary', 'max_salary' and 'salary_range' as there are no true limit for views and applications. However, as salary is paid in different periods, I will be using data column called 'salary_range'.  I will also be utilising the threshold of $\pm$ 3 to identify the Outliers. In using this new data column, a few assumptions will be used.</p>
<br/>
<p><strong><u>Assumptions</u></strong></p>
<ul>
    <li>There is a total of 40 hours of work done per week.</li>
    <li>There is a total of 52 weeks in a year.</li>
    <li>Empirical Rule applies.</li>
</ul>
<br/>
<p>With the assumptions in place, I will proceed to make the new data columns and identify the outliers within the data column.</p>

In [170]:
# ----------- Calculating Z-Score's Variables ----------- #
mean_max = merged_df['salary_range'].mean()
std_max = merged_df['salary_range'].std()

# ----------- Calculating Z-Score ----------- #
merged_df['z_score_salary_range'] = (merged_df['salary_range'] - mean_max) / std_max

# ----------- Identifying Outliers ----------- #
salary_range_outliers_z = merged_df[merged_df['z_score_salary_range'].abs() > 3]

# ----------- Dropping Data Column After Usage ----------- #
merged_df = merged_df.drop(columns=['z_score_salary_range'])

# ----------- Displaying Results ----------- #
print(f"Number of Outliers in salary_range_normalised: {len(salary_range_outliers_z)}")
salary_range_outliers_z

Number of Outliers in salary_range_normalised: 73


,company_name,title,max_salary,pay_period,location,views,min_salary,formatted_work_type,applies,remote_allowed,application_type,expiry,formatted_experience_level,listed_time,max_salary_normalised,min_salary_normalised,postingDuration,salary_range,apply_view_ratio,z_score_salary_range
job_id,,,,,,,,,,,,,,,,,,,,
3886471794,Eastridge Workforce Solutions,Intellectual Property Associate (246215),275000,Hourly,"Dallas, TX",4,240000,Full-time,0,0,SimpleOnsiteApply,2024-05-09 04:46:40,Entry level,2024-04-09 05:20:00,572000000,499200000,29,72800000,0.000000,114.277847
3889115413,Nike,"Store Associate, Seas - Foxwoods",12518,Hourly,"Mashantucket, CT",5,11268,Full-time,0,0,OffsiteApply,2024-05-09 04:46:40,Entry level,2024-04-09 05:20:00,26037440,23437440,29,2600000,0.000000,4.008901
3889116357,Nike,"Store Associate, FT - Vineland",25012,Hourly,"Orlando, FL",7,22512,Full-time,0,0,OffsiteApply,2024-05-09 04:46:40,Entry level,2024-04-09 05:20:00,52024960,46824960,29,5200000,0.000000,8.092936
3889117075,Nike,"Store Associate, PT - Finger Lakes",37506,Hourly,"Waterloo, NY",6,33756,Part-time,0,0,OffsiteApply,2024-05-09 04:46:40,Entry level,2024-04-09 05:20:00,78012480,70212480,29,7800000,0.000000,12.176971
3889117757,The Hillman Group,Full Time Field Service Technician – Wilmingto...,50000,Hourly,"Wilmington, NC",8,45000,Full-time,0,0,OffsiteApply,2024-05-09 04:46:40,Entry level,2024-04-09 05:20:00,104000000,93600000,29,10400000,0.000000,16.261006
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
3902879426,Adams & Martin Group,Trial Paralegal - Hybrid in OC or LA!,100000,Hourly,"Los Angeles County, CA",2,80000,Full-time,0,0,SimpleOnsiteApply,2024-05-19 23:06:40,Mid-Senior level,2024-04-19 23:40:00,208000000,166400000,29,41600000,0.000000,65.269427
3904906761,Unknown,Content Writer,40046666,Yearly,"Virginia, United States",70,28373333,Internship,20,1,SimpleOnsiteApply,2024-05-18 13:46:40,Mid-Senior level,2024-04-18 14:20:00,40046666,28373333,29,11673333,0.285714,18.261135
3904911266,Independent Sports & Entertainment (ISE),Client Relations Manager,80023333,Yearly,"Chicago, IL",22,56686666,Full-time,5,0,ComplexOnsiteApply,2024-10-15 13:46:40,Mid-Senior level,2024-04-18 14:20:00,80023333,56686666,179,23336667,0.227273,36.581699


<p>With reference to the output of the dataframe, I am able to see that the Outlier has been identified successfully.</p>

<h4>3.9.2.4 Removing Identified Outliers Using Z-Score Method</h4>

<p>After identifying some Outliers within the dataframe using the Z-Score method, I will now proceed to remove them from the dataframe. However, I will be using a separate dataframe so as to allow for comparison afterwards.</p>

In [174]:
# ----------- Copy DataFrame for Outlier Removal ----------- #
zScoreDf = merged_df.copy()

# ----------- Define Function to Remove Outliers Using Z-Score ----------- #
def remove_outliers_zscore(df, column, threshold = 3):
    mean = df[column].mean()
    std = df[column].std()
    
    # ----------- Calculating Z-Score ----------- #
    df['z_score'] = (df[column] - mean) / std
    
    # ----------- Removing Values Exceeding Threshold ----------- #
    df_no_outliers = df[df['z_score'].abs() <= threshold]
    
    print(f"Outliers removed for '{column}': {len(df) - len(df_no_outliers)} rows")
    return df_no_outliers.drop(columns=['z_score'])

# ----------- Apply Function for All Three Columns ----------- #
zScoreDf = remove_outliers_zscore(zScoreDf, 'min_salary_normalised')
zScoreDf = remove_outliers_zscore(zScoreDf, 'max_salary_normalised')
zScoreDf = remove_outliers_zscore(zScoreDf, 'salary_range')

# ----------- Display Results ----------- #
print(f"Rows after outlier removal: {len(zScoreDf)}")
zScoreDf.head(5)

Outliers removed for 'min_salary_normalised': 75 rows
Outliers removed for 'max_salary_normalised': 1022 rows
Outliers removed for 'salary_range': 2148 rows
Rows after outlier removal: 120527


,company_name,title,max_salary,pay_period,location,views,min_salary,formatted_work_type,applies,remote_allowed,application_type,expiry,formatted_experience_level,listed_time,max_salary_normalised,min_salary_normalised,postingDuration,salary_range,apply_view_ratio
job_id,,,,,,,,,,,,,,,,,,,
3904995353,Sherwin-Williams,"Sales Representative, Residential Repaint",3366,Biweekly,"Hammonton, NJ",4,2245,Full-time,0,0,OffsiteApply,2024-05-18 16:33:20,Entry level,2024-04-18 17:06:40,87516,58370,29,29146,0.000000
3895598632,Sherwin-Williams,Assistant Manager Trainee,2057,Biweekly,"Sioux Falls, SD",7,2057,Full-time,0,0,OffsiteApply,2024-05-12 05:00:00,Internship,2024-04-12 05:33:20,53482,53482,29,0,0.000000
3904386637,Sherwin-Williams,Assistant Manager Trainee,2134,Biweekly,"Orange, CA",7,2134,Full-time,0,0,OffsiteApply,2024-05-17 23:53:20,Internship,2024-04-17 21:40:00,55484,55484,30,0,0.000000
3863624365,Conquer Employment,Technical Planner,28,Hourly,"Terrell, TX",15,24,Full-time,0,0,ComplexOnsiteApply,2024-05-05 20:13:20,Associate,2024-04-05 20:46:40,58240,49920,29,8320,0.000000
3884430611,Planet Interactive,Digital Designer,40,Hourly,"Bolingbrook, IL",29,35,Contract,4,0,ComplexOnsiteApply,2024-05-05 20:13:20,Associate,2024-04-05 20:46:40,83200,72800,29,10400,0.137931


<p>With reference to the output of the dataframe above, I am able to see that the removal of Outliers have been conducted successfully and I am able to perform comparisiom between the two Outliers Removal Operation.</p>

<hr/>
<h3>2.9.3 Comparison of Outlier Identification & Removal</h3>

<p>After conducting Outlier Identification & Removal using both the IQR method and Z-Score method, I will now compare which method is most suitable to be used and will subsequently use that dataframe for all other operations and Data Analysis.</p>

<hr/>
<h4>2.9.3.1 Description of IQR Method's Dataframe</h4>

<p>To analyse the remaining values for the 'min_salary_normalised', 'max_salary_normalised' and 'salary_range', I will be using the descirbe method on the dataframe which have the Outliers removed. This will allow me to see if there is still a sufficient amount of varying data within the dataframe.</p>

In [180]:
# ----------- Describe Method of IQRdf ----------- #
IQRdf[['min_salary_normalised', 'max_salary_normalised', 'salary_range']].describe()

,min_salary_normalised,max_salary_normalised,salary_range
count,112696.000000,112696.000000,112696.000000
mean,82684.301280,110940.556133,28256.254854
std,33044.907334,44304.099748,19226.453740
min,1.000000,1.000000,0.000000
25%,58846.750000,78333.000000,13334.000000
50%,78750.000000,106250.000000,23334.000000
75%,103000.000000,140000.000000,39785.000000
max,181881.000000,252333.000000,85000.000000


<p>With reference to the dataframe of the IQR Method above. I am able to observe the following.</p>
<ul>
    <li>There are 112696 Data Rows Left.</li>
    <li>There is still a healthy amount of data variety left.</li>
</ul>
<br/>
<p>With the observations listed, I will move on to inspect the dataframe without Outliers produced by using the Z-Score method.</p>

<hr/>
<h4>2.9.3.2 Description of Z-Score Method's Dataframe</h4>

<p>To analyse the remaining values for the 'min_salary_normalised', 'max_salary_normalised' and 'salary_range', I will be using the descirbe method on the dataframe which have the Outliers removed. This will allow me to see if there is still a sufficient amount of varying data within the dataframe.</p>

In [184]:
# ----------- Describe Method of z_filtered_df ----------- #
zScoreDf[['min_salary_normalised', 'max_salary_normalised', 'salary_range']].describe()

,min_salary_normalised,max_salary_normalised,salary_range
count,120527.000000,120527.000000,120527.000000
mean,86961.540252,118751.919338,31790.379085
std,38903.629858,53433.127356,24136.359517
min,1.000000,1.000000,0.000000
25%,60000.000000,80000.000000,14286.000000
50%,80521.000000,110296.000000,25000.000000
75%,107263.500000,149800.000000,43200.000000
max,365533.000000,383333.000000,123534.000000


<p>With reference to the dataframe above, I am able to make the following observations.</p>
<ul>
    <li>There are a total of 120527 remaining data rows.</li>
    <li>There is a variety in data for all three data columns.</li>
</ul>
<p>With the observations stated, I will now move on to the conclusion.</p>

<h3>2.9.4 Conclusion of Method To Utilise</h3>

<p>After completing the identification and removal of Outliers using both the IQR Method and Z-Score Method, and also taking into account the results obtained from the describe method on each of the dataframe, I have come to a conclusion that the <strong><u>IQR Method</u></strong> should be used as it provides a more robust data with higher data rows and data variety. The usage of IQR Method is also safer as we are unsure if the Empirical Rule applies.</p>

In [188]:
# ----------- Making Main Dataframe A Copy of Z-Score Dataframe ----------- #
merged_df = IQRdf.copy()

# ----------- Displaying Dataframe ----------- #
merged_df.head(5)

,company_name,title,max_salary,pay_period,location,views,min_salary,formatted_work_type,applies,remote_allowed,application_type,expiry,formatted_experience_level,listed_time,max_salary_normalised,min_salary_normalised,postingDuration,salary_range,apply_view_ratio
job_id,,,,,,,,,,,,,,,,,,,
3904995353,Sherwin-Williams,"Sales Representative, Residential Repaint",3366,Biweekly,"Hammonton, NJ",4,2245,Full-time,0,0,OffsiteApply,2024-05-18 16:33:20,Entry level,2024-04-18 17:06:40,87516,58370,29,29146,0.000000
3895598632,Sherwin-Williams,Assistant Manager Trainee,2057,Biweekly,"Sioux Falls, SD",7,2057,Full-time,0,0,OffsiteApply,2024-05-12 05:00:00,Internship,2024-04-12 05:33:20,53482,53482,29,0,0.000000
3904386637,Sherwin-Williams,Assistant Manager Trainee,2134,Biweekly,"Orange, CA",7,2134,Full-time,0,0,OffsiteApply,2024-05-17 23:53:20,Internship,2024-04-17 21:40:00,55484,55484,30,0,0.000000
3863624365,Conquer Employment,Technical Planner,28,Hourly,"Terrell, TX",15,24,Full-time,0,0,ComplexOnsiteApply,2024-05-05 20:13:20,Associate,2024-04-05 20:46:40,58240,49920,29,8320,0.000000
3884430611,Planet Interactive,Digital Designer,40,Hourly,"Bolingbrook, IL",29,35,Contract,4,0,ComplexOnsiteApply,2024-05-05 20:13:20,Associate,2024-04-05 20:46:40,83200,72800,29,10400,0.137931


<p>With the output of the dataframe above, I am able to verify that the main dataframe have been removed of Outliers using the IQR method.</p>

<hr/>
<h2>2.10 Validating Numerical Data</h2>

<p>To ensure all Numerical Data are either '0' or a positive value, I will need to validate each data columns. Should there be a value smaller than 0 (in the negatives), I will replace the value with a '0'.</p>

In [192]:
# ----------- Identifying Invalid Values ----------- #
invalidValues = merged_df[['max_salary_normalised', 'min_salary_normalised', 'views', 'applies']].lt(0).sum()

# ----------- Storing in DataFrame ----------- #
invalidValuesDf = pd.DataFrame({'Negative Values': invalidValues.values}, index = invalidValues.index)

# ----------- Function for Colour Coding ----------- #
def colour_code(val):
    if val > 0:
        return 'background-color: red; color: white'
    else:
        return 'background-color: green; color: white'

# ----------- Apply Colour Coding ----------- #
invalidValuesDf = invalidValuesDf.style.map(colour_code, subset=['Negative Values'])

# ----------- Display the Styled DataFrame ----------- #
invalidValuesDf

,Negative Values
max_salary_normalised,0
min_salary_normalised,0
views,0
applies,0


<p>With reference to the dataframe above, I am able to verify that there is no invalid numerical data within the dataframe. Therefore, there is no need to insert a placeholder value of 0 for any of the numerical data.</p>